examples/step-71/step-71.cc

```diff
@@ -0,0 +1,937 @@
+               std::sin(PI * p[1]);
+      }
+    };
+  } // namespace ExactSolution
+
+
+
+  // @sect3{The main class}
+  //
+  // The following is the principal class of this tutorial program. It has
+  // the structure of many of the other tutorial programs and there should
+  // really be nothing particularly surprising about its contents or
+  // the constructor that follows it.
+  template <int dim>
+  class BiharmonicProblem
+  {
+  public:
+    BiharmonicProblem(const unsigned int fe_degree);
+
+    void run();
+
+  private:
+    void make_grid();
+    void setup_system();
+    void assemble_system();
+    void solve();
+    void compute_errors();
+    void output_results(const unsigned int iteration) const;
+
+    Triangulation<dim> triangulation;
+
+    MappingQ<dim> mapping;
+
+    FE_Q<dim>                 fe;
+    DoFHandler<dim>           dof_handler;
+    AffineConstraints<double> constraints;
+
+    SparsityPattern      sparsity_pattern;
+    SparseMatrix<double> system_matrix;
+
+    Vector<double> solution;
+    Vector<double> system_rhs;
+  };
+
+
+
+  template <int dim>
+  BiharmonicProblem<dim>::BiharmonicProblem(const unsigned int fe_degree)
+    : mapping(1)
+    , fe(fe_degree)
+    , dof_handler(triangulation)
+  {}
+
+
+
+  // Next up are the functions that create the initial mesh (a once refined
+  // unit square) and set up the constraints, vectors, and matrices on
+  // each mesh. Again, both of these are essentially unchanged from many
+  // previous tutorial programs.
+  template <int dim>
+  void BiharmonicProblem<dim>::make_grid()
+  {
+    GridGenerator::hyper_cube(triangulation, 0., 1.);
+    triangulation.refine_global(1);
+
+    std::cout << "Number of active cells: " << triangulation.n_active_cells()
+              << std::endl
+              << "Total number of cells: " << triangulation.n_cells()
+              << std::endl;
+  }
+
+
+
+  template <int dim>
+  void BiharmonicProblem<dim>::setup_system()
+  {
+    dof_handler.distribute_dofs(fe);
+
+    std::cout << "   Number of degrees of freedom: " << dof_handler.n_dofs()
+              << std::endl;
+
+    constraints.clear();
+    DoFTools::make_hanging_node_constraints(dof_handler, constraints);
+
+    VectorTools::interpolate_boundary_values(dof_handler,
+                                             0,
+                                             ExactSolution::Solution<dim>(),
+                                             constraints);
+    constraints.close();
+
+
+    DynamicSparsityPattern c_sparsity(dof_handler.n_dofs());
+    DoFTools::make_flux_sparsity_pattern(dof_handler,
+                                         c_sparsity,
+                                         constraints,
+                                         true);
+    sparsity_pattern.copy_from(c_sparsity);
+    system_matrix.reinit(sparsity_pattern);
+
+    solution.reinit(dof_handler.n_dofs());
+    system_rhs.reinit(dof_handler.n_dofs());
```

Comment:
Nope :-) One issue per tutorial program, and here the issue is how to solve the biharmonic equation, not the biharmonic equation in parallel :-)

Reply:
As far a I see it, an extension to a parallel version would be like an additional max. 10 lines and wouldn't add really additional complexity (MPI is since recently a basic technique). But I guess I cannot convince anyone...